compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/ROCDLPrefetchSharedMemoryCopy.cpp

```diff
@@ -301,105 +234,292 @@ class LoopPrefetcher {
     return success();
   }
 
-  /// Creates all read stage ops for a loop iteration with |rewriter| and maps
-  /// the original loop induction variable to |iv| in |mapping|.
-  void emitRead(IRMapping &mapping, RewriterBase &rewriter, Value iv) {
-    // Map the original loop induction variable to |iv| for later op rewrites.
-    mapping.map(forOp.getInductionVar(), iv);
+private:
+  // The original scf.for loop to prefetch shared memory copy from.
+  scf::ForOp forOp;
+  // Original static loop range and step.
+  int64_t lb, ub, step;
+};
+
+} // namespace
 
-    for (Operation *op : readStage) {
-      rewriter.clone(*op, mapping);
+// Removes all barrier operations from the loop body.
+// Original barriers are designed for the unpipelined loop structure and will
+// be replaced with barriers appropriate for the pipelined structure.
+static void removeBarriers(scf::ForOp forOp) {
+  SmallVector<Operation *> toErase;
+  forOp.getBody()->walk([&](Operation *op) {
+    if (isa<gpu::BarrierOp, amdgpu::SchedBarrierOp>(op)) {
+      toErase.push_back(op);
     }
+  });
+  for (Operation *op : toErase) {
+    op->erase();
   }
+}
 
-  /// Creates all write stage ops for a loop iteration with |rewriter| and maps
-  /// the original loop induction variable to |iv| in |mapping|.
-  void emitWrite(IRMapping &mapping, RewriterBase &rewriter, Value iv) {
-    // Map the original loop induction variable to |iv| for later op rewrites.
-    mapping.map(forOp.getInductionVar(), iv);
+// Populates the opToStage map by assigning stages to operations based on
+// prefetcher groups and number of stages.
+static void
+populateOpToStageMap(LoopPrefetcher &prefetcher, scf::ForOp forOp,
+                     unsigned numStages,
+                     llvm::DenseMap<Operation *, unsigned> &opToStage) {
+  auto assignOp = [&](Operation *op, unsigned stage) {
+    if (!op || isa<scf::YieldOp>(op))
+      return;
+    opToStage[op] = stage;
+  };
 
-    for (Operation *op : writeStage) {
-      rewriter.clone(*op, mapping);
+  if (numStages == 1) {
+    // Single-stage pipelining: all ops in stage 0.
+    for (Operation &op : forOp.getBody()->without_terminator()) {
+      assignOp(&op, /*stage=*/0);
```

Comment:
What is the point of everything getting stage 0? Isn't that a no-op?

Reply:
Yes, this is more meant for completeness so we can generate the non-pipelined code if user gives a numStage = 0 to this pass. 

You are also right that plumb the `numStages == 1` case isn't that useful. I didn't pay very close attention to numStages since I was planning to use a future PR to just allow this to be configurable. 

For this PR, I can check for numStages earlier and bail from this pass when `numStages <= 1`.